From 63165d3d6cb892321b3d1a08c5e118abcc8d5185 Mon Sep 17 00:00:00 2001
From: Pavel Grafov <pgrafov@google.com>
Date: Wed, 29 Dec 2021 15:09:38 +0000
Subject: [PATCH] [Q_asb_2022-03] Make ManagedProvisioning hide non-system
 overlays

This is to prevent tapjacking where an app obscures the
provisioning activity and tricks the user into inflating a work
profile without consent.

Bug: 209611539
Test: manual
Change-Id: I4bda57bd80f34402e0a63554e719dceda60a296f
Merged-In: I4bda57bd80f34402e0a63554e719dceda60a296f
(cherry picked from commit f0908ba3ccf4b294be011cb3cb3441b34144f06e)
Merged-In: I4bda57bd80f34402e0a63554e719dceda60a296f
---
 AndroidManifest.xml                                           | 1 +
 .../managedprovisioning/common/SetupLayoutActivity.java       | 4 ++++
 2 files changed, 5 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 0a2f82d6..40abb9be 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -34,6 +34,7 @@
     <uses-permission android:name="android.permission.BIND_DEVICE_ADMIN" />
     <uses-permission android:name="android.permission.MANAGE_DEVICE_ADMINS" />
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
+    <uses-permission android:name="android.permission.HIDE_NON_SYSTEM_OVERLAY_WINDOWS"/>
 
     <uses-permission android:name="android.permission.CHANGE_COMPONENT_ENABLED_STATE" />
     <uses-permission android:name="android.permission.CHANGE_CONFIGURATION" />
diff --git a/src/com/android/managedprovisioning/common/SetupLayoutActivity.java b/src/com/android/managedprovisioning/common/SetupLayoutActivity.java
index 6ae59535..7e63774f 100644
--- a/src/com/android/managedprovisioning/common/SetupLayoutActivity.java
+++ b/src/com/android/managedprovisioning/common/SetupLayoutActivity.java
@@ -17,6 +17,7 @@
 package com.android.managedprovisioning.common;
 
 import static android.view.View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
+import static android.view.WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS;
 
 import static com.android.internal.logging.nano.MetricsProto.MetricsEvent.VIEW_UNKNOWN;
 
@@ -57,6 +58,9 @@ public abstract class SetupLayoutActivity extends Activity {
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+
+        getWindow().addSystemFlags(SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS);
+
         mTimeLogger = new TimeLogger(this, getMetricsCategory());
         mTimeLogger.start();
 
-- 
2.35.1

